clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -1985,47 +1985,63 @@ inlineHO _ e@(App _ _)
 inlineHO _ e = return e
 {-# SCC inlineHO #-}
 
--- | Simplified CSE, only works on let-bindings, works from top to bottom
+-- | Simplified CSE, only works on let-bindings, does an inverse topological
+-- sort of the let-bindings and then works from top to bottom
+--
+-- XXX: Check whether inverse top-sort followed by single traversal removes as
+-- many binders as the previous "apply-until-fixpoint" approach in the presence
+-- of recursive groups in the let-bindings. If not but just for checking whether
+-- changes to transformation affect the eventual size of the circuit, it would
+-- be really helpful if we tracked circuit size in the regression/test suite.
+-- On the two examples that were tested, Reducer and PipelinesViaFolds, this new
+-- version of CSE removed the same amount of let-binders.
 simpleCSE :: HasCallStack => NormRewrite
-simpleCSE (TransformContext is0 _) e@(Letrec binders body) = do
-  let is1 = extendInScopeSetList is0 (map fst binders)
-  let (reducedBindings,body') = reduceBindersFix is1 binders body
-  if length binders /= length reducedBindings
-     then changed (Letrec reducedBindings body')
-     else return e
+simpleCSE (TransformContext is0 _) (inverseTopSortLetBindings -> Letrec bndrs body) = do
+  let is1 = extendInScopeSetList is0 (map fst bndrs)
+  (subst,bndrs1) <- reduceBinders (mkSubst is1) [] bndrs
+  -- TODO: check whether a substitution over the body is enough, the reason I'm
+  -- doing a substitution over the the binders as well is that I don't know in
+  -- what order a recursive group shows up in a inverse topological sort.
+  -- Depending on the order and forgetting to apply the substitution over the
+  -- let-bindings might lead to the introduction of free variables.
```

Comment:
Did you check this `TODO`, or is it for another time?

Reply:
Another time